### DISCUSSION TITLE
More Hydra-zen examples

### DISCUSSION THREAD

**Discussion by francisco-camargo** [mit-ll-responsible-ai/hydra-zen#397] [General]
Hello all,

I am still trying to get my head around what needs to be done on top of writing the "usual" ML code to then be able to use Hydra-zen to track experiments. I have been going through the documentation but not being familiar with PyTorch, the jump from the `game_library.py` example to the PyTorch Lightning one is pretty drastic. In particular, are there any examples using something like `sklearn`, or simpler ML examples in general?

In terms of understanding the PyTorch examples, I get lost trying to understand `UniversalFuncModule`; it's purpose, how it fits into the "usual" ML pipeline, the Hydra-zen pipeline, and what analogous code would be needed in, for example, an sklearn example.

Thanks!
Francisco

**Comment by jgbos** [comment]
@francisco-camargo we completely agree with your sentiment and our docs is one of the things we really want to improve near term.  I'm wondering if you have a particular use case or example that would help you understand how to use hydra-zen with PyTorch?

**Comment by francisco-camargo** [comment]
Hi Justin,

To be honest, the PyTorch example is just not relevant to me at the moment. Don't get me wrong, I am really glad that it is there, I just won't be able to derive much value from it at the moment. My suggestion for simpler examples comes from the assumption that many others will also feel like the docs take a big step in sophistication and thus you'll lose the interest of many readers (ie. I don't want to also learn PyTorch today, I just want to learn hydra-zen).

I am still trying to understand the basic building blocks of hydra-zen that generally get used. I will try to spend more time the next couple days rereading and trying to understand the examples in the docs and I may be able to give more constructive feedback. At the moment I would say that coming at this as someone that is not a software engineer but am interested because I want to help and encourage my engineers to have more robust experimental setups up, the docs are just not that easy to digest. For example, I am able to follow the logic of the code presented, but when I finish reading each section I don't have the sense that I would be able to easily apply what I just read to my own problems.

I suggest having `sk-learn` examples both because of how ubiquitous it is pedagogically, but also because it would (I hope) allow for examples where hydra-zen could make it easy to swap between different learners (linear regression, decision tree, SVM, etc.), each with their own set of tunable hyperparameters (ie. setting up a set of GridSearch's), and do so outside of the code. Something like this would personally make it a much easier sell if in the future I wanted to advocate for others to adopt the platform.

For some context; I am involved in developing some data science curriculum for novices and from the content I see out there, the experimental setup side of things is always overlooked. So to that end I am very interested to not only use a tool such as this myself but find a tool that could be taught easily.

If you can use some help with this, time provided, I'd be happy to contribute where I can.

**Reply by jgbos** [reply]
Thanks for this feedback, we really appreciate it.  I definitely agree with this sentiment and agree we need more examples, especially ones that don't require learning an additional framework (like PyTorch and Lightning).  You may be interested in discussions we started today in:

#389 
#395 

I plan to incorporate those discussions into our docs.

I really like the idea of using `sklearn` because they have many different examples and it is easy to use out of the box.  Given that, if there are any particular examples you'd like to see in hydra-zen feel free to let us know. Or start a discussion for help on solving a problem.  

Also, that's great that you're looking to utilize hydra-zen to help teach data science.  I think @rsokl is involved in similar activities.  If you have anything to share or want us to take a look at the code feel free to reach out.

**Reply by rsokl** [reply]
@francisco-camargo thanks so much for your feedback. As an initial measure, we have [added some new How-To sections](https://mit-ll-responsible-ai.github.io/hydra-zen/changes.html#documentation-2023-01-22), including one involving scikit-learn.

> So to that end I am very interested to not only use a tool such as this myself but find a tool that could be taught easily.
> If you can use some help with this, time provided, I'd be happy to contribute where I can.

This would be much appreciated! We leverage the [Diataxis framework for technical authoring](https://diataxis.fr/) to organize/motivate the different sections of our documentation. I think we need more/better Explanation sections, including a general overview of Hydra and hydra-zen. Please feel to share any ideas that you have for additional materials for our docs. And if you do have any interest in contributing, I would be more than happy to assist.

**Reply by francisco-camargo** [reply]
Hey guys, 

Just skimmed through the new section and it looks great, I'm eager to dive into it soon as I can! This is the first I hear of Diataxis but I'll read up on it. I am certainly interested in contributing however I can, this is a really promising and interesting project! I am fairly time constrained at the moment between full-time work and part-time grad school but it's definitely been on my mind to try to incorporate hydra-zen into one or both of these. Pie-in-the-sky, this year I would love to work on an introductory ML curriculum that includes content on "experimentation" and so hydra/hydra-zen seems like a natural choices.
